### PR TITLE
Add wharel method to query with a block

### PR DESCRIPTION
**Pull request by brunvez**
### Summary
**TL;DR**: exposing a subset of Arel and using blocks to build queries that need to be done through prepared statements right now.

This PR is a follow-up to https://github.com/rails/rails/pull/39445 which aims to fix some of the things pointed out and give a fresh starting point without so many comments.

As a recap this PR tries to address are the two following types of queries:

```sql
SELECT posts.* FROM posts WHERE posts.views > 100 OR posts.title LIKE '%Rails%'
``` 

```sql
SELECT `authors`.* FROM `authors` INNER JOIN `posts` ON `posts`.`author_id` = `authors`.`id` WHERE `posts`.`tags_count` > 10
```

Which are solved right now using something similar to prepared statements like `Post.where("views > ? OR title LIKE ?", 100, "'%Rails%'")`. This syntax has many drawbacks:
- First of all ambiguity: this is a classic in many applications I've seen
```ruby
Post.where('created_at > ?', 1.month.ago).joins(:comments)
# run this query and you'll get an ActiveRecord::StatementInvalid (PG::AmbiguousColumn: ERROR:  column reference "created_at" is ambiguous) exception
```
The solution is to fully qualify the columns, but that can be very verbose if you have big table names
- DB specific syntax: the where blocks describe it perfectly
> While this provides the most flexibility, you can also unintentionally introduce dependencies on the underlying database
- Possible SQL injections: where an unaware developer can do something like `Post.where("title LIKE '%#{params[:query]}%'")`

The solution I've generally reach for is Arel, which allows building complex queries without having to write raw SQL or prepared statements. The issue there is that Arel is still a private API so I don't want to rely too much on it but mostly that it just doesn't feel like the Rails way.

My idea of how querying can be enhanced is simple: use a block to build the query just as if you were using `#select` or `#map`. From the previous thread, it seems like an acceptable name would be `#wharel` (I'd still rather use `#where` with a block) and could be considered as an advanced feature. Also, since there was some controversy around exposing the whole private interface I think we could just release a subset of it that allows controlling the implementation details behind it (it's a simple wrapper right now but it can be implemented in any sort of way). This can easily be expanded to fit other cases like `avg`, `min`, `max`, etc.

Replicating the queries from before it would look like
```ruby
Post.wharel { |post| post.views.gt(100).or(post.title.matches("%Rails%") }
```

And chaining with joined tables would look like
```ruby
Author.joins(:posts).wharel { |author| author.posts.tags_count.gt(10) }
```

Sorry for the length but I wanted to provide proper context as to why I'd like this change to move forward. I've also been told that maybe it'd be good if @dhh could weigh in since last time the discussion went off on a tangent.

### CODE REVIEW

**Comment by brunvez** [comment]
Hey @dhh sorry to bother, I know it's been crazy lately, but could you give me an opinion on this?

**Comment by dhh** [comment]
Thanks for exploring this @brunez. I’ve toyed for years with different query syntaxes, but have never found one that beat what we have already, and I don’t think this does it either.

The basic a/b on your examples:

```ruby
Post.wharel { |post| post.views.gt(100).or(post.title.matches("%Rails%") }

Post.where("views > ? OR title LIKE ?", 100, "'%Rails%'")


Author.joins(:posts).wharel { |author| author.posts.tags_count.gt(10) }

Author.joins(:posts).where(“posts.tag_count > ?”, 10)
```

I don’t find the argument that this might lead to sql injection compelling, since we’d still have the existing API, and newcomers are unlikely to learn a custom query syntax if they weren’t going to follow the basic approach re sql injection.

Same with the ability to use DB specific calls. IMO a sharp knives feature, and not commonly seen cutting fingers off either.

The argument I find most compelling is the disambiguation one. Not so much when the whole query is constructed as a single expression, but when a scope is passed around and being built up. I’ve occasionally seen that.

But the remedy is pretty clear, and it’s a fail fast scenario. So criticality is low.

Contrast that with the cost of introducing a new query interface. Two ways of doing the same thing that largely comes down to preferences, and where mixing and matching is going to be ugh.

So I remain unconvinced that a convenient API along these lines and using these examples is either warranted or good. HOWEVER I remain interested in the possibility that we could find a query where it really is substantially better, and that this query would be rare enough to tolerate an approach with a high dose of syntactic vinegar to discourage common usage.

I could see something like .where_query_relation or something similarly long and cumbersome (preferably even longer and even more cumbersome!) is warranted if we can uncover a set of queries that are substantially superior with Arel.

But first we have to find those real-life queries where the A/B is unequivocally in favor of using arel.
